core/state_processor.go

```diff
@@ -131,17 +135,20 @@ func (p *StateProcessor) Process(block *types.Block, statedb *state.StateDB, cfg
 // and uses the input parameters for its environment similar to ApplyTransaction. However,
 // this method takes an already created EVM instance as input.
 func ApplyTransactionWithEVM(msg *Message, config *params.ChainConfig, gp *GasPool, statedb *state.StateDB, blockNumber *big.Int, blockHash common.Hash, tx *types.Transaction, usedGas *uint64, evm *vm.EVM) (receipt *types.Receipt, err error) {
-	if evm.Config.Tracer != nil && evm.Config.Tracer.OnTxStart != nil {
-		evm.Config.Tracer.OnTxStart(evm.GetVMContext(), tx, msg.From)
-		if evm.Config.Tracer.OnTxEnd != nil {
-			defer func() {
-				evm.Config.Tracer.OnTxEnd(receipt, err)
-			}()
+	var tracingStateDB = vm.StateDB(statedb)
```

Comment:
Why don't we allocate a hooked statedb and reuse it for all transactions?

Reply:
Initially, I created the hooked statedb higher up in the call hierarchy, but the is that at this location, we still need the `*state.StateDB`, to perform the `IntermediateRoot`. 

The higher we place the instantiation of hooked, we're forced to either cram more methods into `vm.StateDB` interface (which I did initially), _or_ we need to pass both the `*state.StateDB` struct and the `vm.StateDB` interface side by side. 

I'll take another look at what to do. Reusing is nice.

Reply:
So, this method needs the `*state.StateDB` here, in order to create the receipt before return. We _could_ create it in the caller, and pass both of them... but there are three separate callsites which call in here, and IMO doing this in those three locations is messier than constructing it here. 

It's a pretty small object, shouldn't make that much difference IMO.

Reply:
yeah, gotcha, sgtm